### PR TITLE
Add support for proxied environments

### LINKED ISSUE #1682: [Enhancement] Support http_proxy
* `carthage version`:  0.18.1
* `xcodebuild -version`:  8C1002
* Are you using `--no-build`? No
* Are you using `--no-use-binaries`? No
* Are you using `--use-submodules`? No

Carthage can't use proxy server to download prebuilt binaries.
It's a much needed feature for developers in China.

### PR DESCRIPTION
**Pull request by jakeadams**
Completes PR #1996, which resolves #1682.

### CODE REVIEW

**Comment by blender** [comment]
@jakeadams thanks for the contribution!. Is this an addition to #1996 or can it be merge alone and include #1996 ?

**Comment by jakeadams** [comment]
@blender It can be merged alone. I started from scratch off latest master.

**Comment by blender** [comment]
Summarizing for @mdiep 

### Goal

Support HTTP Proxies. Completes & supersedes #1996 

### Motivation

Not all environments have unrestricted access to HTTP resources.

### Current limitations

No support

### Proposed solution

Grab http_proxy from environment config.

### Example:
```bash
#defined in current process environment
http_proxy=https://github.com:8888
HTTP_PROXY=https://github.com:8888
```

File: Source/CarthageKit/Proxy.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import Foundation
+
+struct Proxy {
+	let connectionProxyDictionary: [AnyHashable: Any]?
+
+	init(environment: [String: String]) {
+		let http = Proxy.makeHttpDictionary(environment)
+		let https = Proxy.makeHttpsDictionary(environment)
```
Comment (mdiep) [inline Source/CarthageKit/Proxy.swift]: `HTTP` should be all-caps in both of these.

File: Source/CarthageKit/Proxy.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		if let port = proxyURL.port {
+			dictionary[kCFNetworkProxiesHTTPSPort] = port
+		}
+
+		return dictionary
+	}
+}
+
+extension Proxy {
+	static let value: Proxy = Proxy(environment: ProcessInfo.processInfo.environment)
```
Comment (mdiep) [inline Source/CarthageKit/Proxy.swift]: I don't love `value` here. Maybe `environment` or `default`?

**Comment by mdiep** [comment]
Thanks for the PR! 🎉 

Overall approach looks great. I love all the tests. Just a couple minor naming issues.

File: Source/CarthageKit/Proxy.swift (hunk 2 above)
Comment (jakeadams) [reply Source/CarthageKit/Proxy.swift]: I wasn't a fan either. I'll go with `default`!

**Comment by jakeadams** [comment]
Along with the name changes requested, I added 1 more test to validate the behavior if a malformed URL is provided.

File: Source/CarthageKit/Proxy.swift [hunk 3]
```diff
@@ -4,10 +4,10 @@ struct Proxy {
 	let connectionProxyDictionary: [AnyHashable: Any]?
 
 	init(environment: [String: String]) {
-		let http = Proxy.makeHttpDictionary(environment)
-		let https = Proxy.makeHttpsDictionary(environment)
+		let HTTP = Proxy.makeHttpDictionary(environment)
+		let HTTPS = Proxy.makeHttpsDictionary(environment)
```
Comment (mdiep) [inline Source/CarthageKit/Proxy.swift]: Oops. I meant this:

```swift
let http = Proxy.makeHTTPDictionary(environment)
```

😅

**Pull request merged**